### ISSUE TITLE
Let's put custom stuff in a custom.el file so that we can ignore it.

### ISSUE THREAD

**Issue by CestDiego** [syl20bnr/spacemacs#1639]

**Comment by justbur** [comment]
I was thinking it might be nice to have a spacemacs directory (default at ~/spacemacs but with a configurable location) with init.el (for .spacemacs), custom.el, and all private layer files. Then it would be easy to keep all of this as a separate repo.

**Comment by CestDiego** [comment]
^ this

**Comment by syl20bnr** [comment]
@justbur This is the responsibility of the user to choose how to manage their dotfiles, you propose one way to do it, that's fine but no everybody will like it. Right now you can already do what you describe.

I want to keep it simple in spacemacs by keeping only one file for configuration, that is the dotfile.

Now we may need to support custom location for customized variable but by default I want to keep them in the dotfile. We can introduce `dotspacemacs-customized-variables-file` variable.

**Comment by syl20bnr** [comment]
> Right now you can already do what you describe.

I mean you can put your layers and doftile and anything else that matters in your dotfile repo and then install them at the required place (if it is even needed) in your necessary install script.

**Comment by justbur** [comment]
@syl20bnr I completely agree that initially with a single .spacemacs file and minimal customizations the current setup is the simplest. 

I think what @CestDiego was getting at and what I was thinking too is that once you start to have more meaningful customizations, .spacemacs grows because of the custom variables, and you need a place to start putting your own layers (which may start out as /private). Then in my opinion the simplest thing to do is to create a spacemacs directory/repo somewhere and hook everything in. It would be even simpler if spacemacs recognized this directory and you didn't have to bother with symlinks.

Your right that the user has the ability to organize things to their liking, and I'm not unhappy with the current design. I just thought my suggestion followed in the spirit of keeping the organization simple (for someone who wants to add custom variables and layers).

This could even start as simply as making .spacemacs a folder with one file in it to start with this file being equivalent to the current .spacemacs, like .spacemacs/init.el or something, and then you could describe how to (for those who wanted to) (i) split off custom variables into custom.el; (ii) add layers to a .spacemacs/layers directory; etc. 

That being said, it was just a thought, nothing more.

**Comment by syl20bnr** [comment]
Private layers does not need to be moved around, a variable in the dotfile points to them. For any other emacs lisp files you want to add you can in `dotspacemacs/init` add them to the load-path variable and then require them, or directly load them with `load-file`.
Moreover splitting the dotfile is not really necessary in spacemacs, you already have the layer concept to split configuration into semantic units.

**Comment by syl20bnr** [comment]
TL;DR I won't create a directory with a single file in it because the user may want to put additional files there. I think this is the responsability of any dotfile engine or custom script the user is using.

**Comment by CestDiego** [comment]
Ye... but @syl20bnr this issue was about putting a custom `custom-file`.  It's really weird when I have to share custom files between windows and linux and I have defaults set for each one that are dependent on OS... for every new `restart` I have to erase the bottom part of my .spacemacs and then when committing new changes I have to avoid comiting the custom part of it. Having the customs.el in a separate file and ignoring it by default would make much of a difference. The important part is...where should it be located?

By default I think it should be in .emacs.d and ignored by git. Thing is ... whenever one want to start from a `cleanslate` one has to always remember to delete that custom.el file.... :( tough stuff ;_;

**Comment by syl20bnr** [comment]
I answered this above:

> Now we may need to support custom location for customized variables but by default I want to keep them in the dotfile. We can introduce dotspacemacs-customized-variables-file variable.

To start in a clean state I rename temporaily my dotfile.

**Comment by CestDiego** [comment]
Also I think that the way of how one `extends` one layer/package is yet not well explained. I mean, what if I want to extend the org config layer? should I use a
`my-org-config` layer? and then
```elisp
(defun my-org-config/init-org ()
    (use-package org
        :config
        ;; My own config goes here....will it override org?
    )
)
```

what's the true difference between

```elisp
(defun my-org-config/post-init-org ()
    ;; (use-package org?)
    ;; Should I put `org` in the list of packages above?
    ;; Can this be inside another use-package?
    ;; should I just eval-after-load everything? D:
)
```

**Comment by syl20bnr** [comment]
This is not yet really documented, it is explained in the release notes I believe.
We can discuss about this in another thread. :-)

**Comment by CestDiego** [comment]
Ok so by default there is no `custom-file-variable` and so it won't be in the `.emacs.d` dir and it's the user's responsibility to keep it in their own repo, and ignore it he wants to...gr8 that sound good :)

**Comment by CestDiego** [comment]
@syl20bnr Ok.... phew I was just using them without knowing then x3

**Comment by syl20bnr** [comment]
The variable is set here: https://github.com/syl20bnr/spacemacs/blob/develop/spacemacs/config.el#L214
So it is not easily overridable for now.

**Comment by CestDiego** [comment]
Oh I see ;_; let's hope til 1.0 :P

**Comment by justbur** [comment]
@syl20bnr Don't be mad, but I had one more idea on this and then I'll shut up. 

What if the design was give priority to the .spacemacs file if it exists, but then search for .spacemacs.d/init.el if it doesn't? Similar to how .emacs and .emacs.d/init.el work. So the current setup would work exactly as it does now, but I could have the option of moving .spacemacs to .spacemacs.d/init.el and having a nicely organized directory with no symlinks to worry about.

**Comment by louy2** [comment]
The point of a separate custom file IMO is to separate Emacs generated custom settings and manual settings. Spacemacs would be all the manual settings we have, and all the Emacs generated custom settings should be user-specific. Why not just throw it into `private` and call it a day? Then all I will worry about is managing `private`.

**Comment by CestDiego** [comment]
with the closing of 1871 I think this is resolved

**Comment by justbur** [comment]
I would still like to do this at least for those using the dotspacemacs directory. I've held off on a pr because I don't want to deal with the change over. there's some work to do if i remember right. It's not handled automatically.

**Comment by SteveJobzniak** [comment]
For those who stumble across this in **insert current year here:**

You cannot and will never be able to set `custom-file` in Spacemacs. If you try to do it, you will break the custom-setting system.

Some dotspacemacs files online try to do it, but all you're doing is screwing everything up.

The code of `~/.emacs.d/core/core-custom-settings.el` literally hardcodes "find the dotspacemacs file" (it calls `dotspacemacs/location` which returns your dotfile) regardless of what we set spacemacs--custom-file to.

It then deletes the custom section from your dotfile and writes the new one there. In fact, spacemacs--custom-file is just a temporary cache location, and it then forcibly merges that into your config file afterwards.

So no. People. Stop trying to affect `custom-file`. It won't work. Pull requests are welcome if you think you can rewrite Spacemacs `core-custom-settings.el`. But until then, you're stuck with custom settings in your dotspacemacs file itself.

**Comment by d12frosted** [comment]
@SteveJobzniak 

To be fair, this particular issue was raised long before Spacemacs [started][1] to inject custom settings into separate function in `.spacemacs` file. Before it this file was used as a `custom-file`, so changing it's value was good enough to get rid of custom settings in the end of `.spacemacs` file. And there is [separate discussion][2] for getting rid of that function.

  [1]: https://github.com/syl20bnr/spacemacs/commit/e699f1816a680c69455273a00c6dceca40796129
  [2]: https://github.com/syl20bnr/spacemacs/issues/7891

**Comment by SteveJobzniak** [comment]
@d12frosted Ahh, so that forced writing is a recent change.! I thought this discussion was about that, and that people had now lived with it for years.

I guess it will end up annoying someone enough that they create a pull request after all. ;-)

**Comment by whirm** [comment]
I can't seem to find `core-custom-settings.el`. Is @SteveJobzniak's comment still valid? 

I just added
```
  (setq custom-file (file-truename (concat dotspacemacs-directory "custom.el")))
  (load custom-file)
```
To my `spacemacs/user-config` and everything seems to be fine.

**Comment by Somelauw** [comment]
@whirm core-custom-settings.el only exists on [develop](https://github.com/syl20bnr/spacemacs/tree/develop) so far. If you're on master, you still have the freedom to put your custom-file elsewhere. Judging from the [recent commit history](https://github.com/syl20bnr/spacemacs/commits/develop/core/core-custom-settings.el), it doesn't seem to have changed since SteveJobzniak's comment.

**Comment by whirm** [comment]
Oh, I see. I would be sad to no longer be able to store my custom settings to a separate file.

Thanks!

**Comment by shajra** [comment]
The conversation above is a bit long, largely because people _really_ want ~/.spacemacs to be the non-custom configuration.  But if we give up on this goal, because of the hardcoding/rigidity discussed above, I think it's easy enough to just use a simple `load-file` call to leave ~/.spacemacs for custom variables, and delegate to another file for static configuration:

```lisp
;; sole contents of ~/.spacemacs
(load-file "~/.spacemacs.static")
```
And then in ~/.spacemacs.static, you can put configuration that Emacs shouldn't touch dynamically.  If there's an easier way now, please let me know.

I don't want to reopen this ticket.  I just wanted to tidy the discussion up for posterity in case someone else stumbled across this ticket as I did.

**Comment by SteveJobzniak** [comment]
@shajra Custom file is possible with latest develop version of Spacemacs, see this example:

https://github.com/SteveJobzniak/spacemaxed/commit/e39f8f5b4747ebf7cb9801fd9b8b13e435acb348

**Comment by eeshugerman** [comment]
the link above is broken. anyone have an example?

**Comment by whirm** [comment]
> the link above is broken. anyone have an example?

Just add this to your `user-init` 
```elisp 
(defun dotspacemacs/user-init ()
[...]
  (customize-set-variable
    'custom-file (file-truename (concat dotspacemacs-directory "custom.el")) "Separate custom file")
  (load custom-file)
[...]
)
```

**Comment by eeshugerman** [comment]
thanks! that does the trick.

**Comment by vainPointer** [comment]
According to above, my config goes as
`
    (setq third-path "~/.emacs.d/third")
    (setq third-packages (directory-files-recursively third-path ".*"))
    (while third-packages (load-file (car third-packages)) (setq third-packages (cdr third-packages)))
`
